src/ai/backend/manager/event_dispatcher/model_serving.py

```diff
@@ -0,0 +1,225 @@
+
+from ai.backend.common.events.model_serving import (
+    ModelServiceStatusEvent,
+    RouteCreatedEvent,
+)
+from ai.backend.common.types import (
+    AgentId,
+    ClusterMode,
+    ImageAlias,
+    ModelServiceStatus,
+    SessionTypes,
+)
+from ai.backend.logging import BraceStyleAdapter
+from ai.backend.manager.errors.exceptions import SessionNotFound
+from ai.backend.manager.registry import AgentRegistry
+
+from ..models.endpoint import EndpointRow
+from ..models.image import ImageIdentifier, ImageRow
+from ..models.keypair import KeyPairRow
+from ..models.routing import RouteStatus, RoutingRow
+from ..models.session import KernelLoadingStrategy, SessionRow
+from ..models.user import UserRow
+from ..models.utils import (
+    ExtendedAsyncSAEngine,
+    execute_with_retry,
+    is_db_retry_error,
+)
+from ..types import UserScope
+from ..utils import query_userinfo
+
+log = BraceStyleAdapter(logging.getLogger(__spec__.name))
+
+
+class ModelServingEventHandler:
+    def __init__(self, registry: AgentRegistry, db: ExtendedAsyncSAEngine) -> None:
+        self._registry = registry
+        self._db = db
+
+    async def _handle_model_service_status_update(
+        self,
+        context: None,
+        source: AgentId,
+        event: ModelServiceStatusEvent,
+    ) -> None:
+        log.info("HANDLE_MODEL_SERVICE_STATUS_UPDATE (source:{}, event:{})", source, event)
+        try:
+            async with self._registry.db.begin_readonly_session() as db_sess:
+                session = await SessionRow.get_session(
+                    db_sess,
+                    event.session_id,
+                    allow_stale=False,
+                    kernel_loading_strategy=KernelLoadingStrategy.MAIN_KERNEL_ONLY,
+                )
+                route = await RoutingRow.get_by_session(db_sess, session.id, load_endpoint=True)
+        except SessionNotFound:
+            return
+        except NoResultFound:
+            return
+
+        async def _update() -> None:
+            async with self._db.begin_session() as db_sess:
+                data: dict[str, Any] = {}
+                match event.new_status:
+                    case ModelServiceStatus.HEALTHY:
+                        data["status"] = RouteStatus.HEALTHY
+                    case ModelServiceStatus.UNHEALTHY:
+                        data["status"] = RouteStatus.UNHEALTHY
+                query = sa.update(RoutingRow).values(data).where(RoutingRow.id == route.id)
+                await db_sess.execute(query)
+
+                query = sa.select(RoutingRow).where(
+                    (RoutingRow.endpoint == route.endpoint)
+                    & (RoutingRow.status == RouteStatus.HEALTHY)
+                )
+                result = await db_sess.execute(query)
+                latest_routes = result.fetchall()
+                latest_routes = await RoutingRow.list(
+                    db_sess, route.endpoint, status_filter=[RouteStatus.HEALTHY]
+                )
```

Comment:
The behavior here is a little strange...

Reply:
hmm... I copied this part but we need to check the code is correct